autocorns/biologist.py

```diff
@@ -98,149 +127,137 @@ def unicorn_metadata(
     dnas: List[Dict[str, Any]],
     block_number: Optional[int] = None,
-    num_workers: int = 1,
-    timeout: float = 30.0,
-    checkpoint_file: Optional[str] = None,
 ) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
     if block_number is None:
         block_number = len(chain) - 1
 
     results: List[Dict[str, Any]] = []
     errors: List[Dict[str, Any]] = []
 
-    dnas_index = {item["token_id"]: item for item in dnas}
-
-    checkpointed_token_ids: Set[int] = set()
-    if checkpoint_file is not None:
-        with open(checkpoint_file, "r") as ifp:
-            for line in ifp:
-                stripped_line = line.strip()
-                if stripped_line:
-                    result = json.loads(stripped_line)
-                    if result["token_id"] not in dnas_index:
-                        continue
-                    dna_result = dnas_index[result["token_id"]]
-                    if (
-                        result.get("num_mythic_body_parts") is not None
-                        and result.get("dna") != "2"
-                        and dna_result["block_number"] <= result["block_number"]
-                    ):
-                        results.append(result)
-                        checkpointed_token_ids.add(result["token_id"])
-
-    submission_progress_bar = tqdm(
-        total=len(dnas) - len(checkpointed_token_ids),
-        desc="Submitting requests for number of unicorn mythic body parts",
-    )
+    tokens_metadata = []
 
     mythic_progress_bar = tqdm(
-        total=len(dnas) - len(checkpointed_token_ids),
+        total=len(dnas),
         desc="Retrieving number of unicorn mythic body parts",
     )
 
     contract = StatsFacet.StatsFacet(contract_address)
 
-    jobs: Dict[Future, int] = {}
-    with ThreadPoolExecutor(max_workers=num_workers) as executor:
-        for item in dnas:
-            if item["token_id"] in checkpointed_token_ids:
+    block_number = dnas[0]["block_number"]
+
+    multicaller = Multicall2.Multicall2(Multicall2_address)
+
+    multicall_method = multicaller.contract.tryAggregate
+
+    dnas_is_present = [
+        dna for dna in dnas if dna["dna"] is not None and dna["dna"] != "None"
+    ]
+
+    for dnas_chunk in [
+        dnas_is_present[i : i + CALL_CHUNK_SIZE]
+        for i in range(0, len(dnas), CALL_CHUNK_SIZE)
+    ]:
+        while True:
+            try:
+
+                send_to_multicall_dnas = [dna["dna"] for dna in dnas_chunk]
+
+                make_multicall_result = make_multicall(
+                    multicall_method,
+                    contract.contract.getUnicornBodyParts,
+                    contract_address,
+                    send_to_multicall_dnas,
+                    block_number=block_number,
+                )
+                tokens_metadata.extend(make_multicall_result)
+                mythic_progress_bar.update(len(send_to_multicall_dnas))
+                break
+            except ValueError:
+                time.sleep(1)
```

Comment:
Retry interval should be a parameter.

Reply:
I would also take number of retries as a parameter.

Reply:
Add latter not used i will remove for now.